src/main.rs

```diff
@@ -0,0 +1,72 @@
+use std::time::Duration;
+
+use clap::Parser;
+
+use nomad_external_dns::{hetzner_dns, nomad};
+use reqwest::Url;
+use tokio::time::sleep;
+
+use nomad_external_dns::config::{Config, DnsProvider};
+use nomad_external_dns::consul::ConsulClient;
+use nomad_external_dns::dns_trait::DnsProviderTrait;
+
+#[tokio::main]
+async fn main() {
+    println!("Starting up Nomad External DNS");
+
+    let config = match Config::try_parse() {
+        Ok(config) => config,
+        Err(e) => {
+            eprintln!("Failed to parse configuration: {}", e);
+            return;
+        }
+    };
+    println!("Configuration parsed successfully");
+
+    // Initialize Consul Client
+    let consul_client = loop {
+        match ConsulClient::new(
+            Url::parse(&config.consul_address).expect("Invalid URL"),
+            config.consul_datacenter.clone(),
+        ) {
+            Ok(client) => break client,
+            Err(e) => {
+                eprintln!("Failed to create Consul client: {}. Retrying in 100ms", e);
+                sleep(Duration::from_millis(100)).await;
+            }
+        }
+    };
+    println!("Consul client created successfully");
+
+    // Acquire Lock
+    if let Err(e) = consul_client.acquire_lock().await {
+        eprintln!("Failed to acquire Consul lock: {}", e);
+        return;
+    }
+    println!("Consul lock acquired successfully");
+
+    let nomad_tag = match nomad::fetch_and_parse_service_tags(&config).await {
+        Ok(tag) => tag,
+        Err(e) => {
+            eprintln!("Failed to fetch Nomad DNS tags: {}", e);
+            return;
+        }
+    };
+    println!("Nomad DNS tags fetched successfully");
+
+    let dns_provider: Box<dyn DnsProviderTrait> = match config.dns_provider {
+        DnsProvider::Hetzner(config) => Box::new(hetzner_dns::HetznerDns { config }),
+    };
+
+    let result = dns_provider.update_or_create_dns_record(&nomad_tag).await;
```

Comment:
As currently implemented this appears to expect to be executed separately for each deployed Nomad job, but what we actually want is to have a single deployed instance that operates on _all_ services registered in Consul.  (And so arguably we should rename this to `consul-external-dns` instead, which conveniently also mitigates the naming conflict with the old project.)

To do this you want to use Consul's `/v1/catalog/services` endpoint.  I'd say that we want to query it with at least a filter of `ServiceKind == ""` to only match "normal" Consul services.  Though as I just wrote in our ticket I think it would make sense for us to mandate a `external-dns.enable=true` tag be set on the services so we can fetch a more limited set of relevant services using a filter like `ServiceKind == "" and ServiceTags contains "external-dns.enable=true"`.

If you have any questions, ping me on Discord.